### PR TITLE
Jam rest of docs

### LINKED ISSUE #102: Document the rest
- Document the rest to Doxy compatible javadoc standard.
- Fix filter.hh not having th  timeout and having the output pipe malformed on the constructor.
- Document the pipe extra carefully, potentially with a usage example.

Filter.hh and the [doxy manual](https://www.doxygen.nl/manual/docblocks.html) may be useful.

### PR DESCRIPTION
**Pull request by Jambls**
closes #102 #103

### CODE REVIEW

**Comment by MHCooke** [comment]
Can we get something on the home page?

File: include/callbacks.hh [hunk 1]
```diff
@@ -37,45 +12,71 @@ int main () {
...
+ * alsa callback namespace
+ */
 namespace alsa_callback {
 /**
  * Alsa Callback Class. This reads Audio from the input hardware device and, using a callback will write these
  * to the correct pipes.
+ *
+ *  Example usage of this functionality:
+ *
+ * ===================================================
```
Comment (MHCooke) [inline include/callbacks.hh:25]: This line of equals puts a horizontal divider in the docs website which doens't look great, can you remove it

File: include/callbacks.hh [hunk 2]
```diff
@@ -37,45 +12,71 @@ int main () {
 #include <thread>
 #include <vector>
 
+/**
+ * alsa callback namespace
+ */
 namespace alsa_callback {
 /**
  * Alsa Callback Class. This reads Audio from the input hardware device and, using a callback will write these
```
Comment (MHCooke) [inline include/callbacks.hh:20]:
```suggestion
 * Alsa Callback Class. This reads Audio from the input hardware device and, using a callback, will write these
```

File: include/callbacks.hh [hunk 3]
```diff
@@ -37,45 +12,71 @@ int main () {
...
+  public:
+    /**
+     * Constructs an ALSA Callback. Initialises the device name.
+     * @param device_name A string describing the ALSA device to be used (e.g. "hw:2,0,0").
+     */
+    acb(const std::string &device_name);
+    /**
+     * Run function. Opens a thread and continuously reads the hardware device input and triggers a callback.
+     */
+    void listen(const std::function<void(const std::vector<int16_t> &, const std::vector<int16_t> &)> &callback);
```
Comment (MHCooke) [inline include/callbacks.hh:56]: Missing param for callback

**Review by bear-in-the-air — APPROVED** [review]

**Comment by MHCooke** [comment]
In alsa_out.hh the constructor tmoutput param should be:

> Time in milliseconds the thread will wait before exiting if no data is provided.

**Comment by MHCooke** [comment]
Amplifier::amplify() is missing the @return description

File: include/filter.hh
Comment (MHCooke) [inline include/filter.hh:1]: Add link to Bernd's github: https://github.com/berndporr/iir1

Comment (MHCooke) [reply include/filter.hh:1]: filter() is missing argument and return docs

File: include/pipe.hh [hunk 4]
```diff
@@ -5,10 +5,46 @@
...
+ *   output.cond.notify_all();
+ * }
+ *
+ * ~~~~~~~~~~~~~~~
+ */
 template <typename T> struct Pipe {
-    std::condition_variable cond;
-    std::mutex cond_m;
-    std::queue<T> queue{};
+    std::condition_variable cond; /**< condition variable for the pipe*/
```
Comment (MHCooke) [inline include/pipe.hh:45]: Condition variable used to wait/notify

File: include/pipe.hh
Comment (MHCooke) [inline include/pipe.hh:1]: Would be nice in the three struct member descriptions were capitalised

**Comment by MHCooke** [comment]
RMS constructor. num samples is repeated

**Review by MHCooke — APPROVED** [review]

**Pull request merged**